### DISCUSSION TITLE
Unable to explicitly set Content-Encoding on API Route response.

### DISCUSSION THREAD

**Discussion by Cybertrance** [vercel/next.js#48827] [Help] [answered]
### Summary

Hello!

Having recently upgraded to v13 from the older 12.3.2, we've hit a roadblock with one of our API endpoints.
Basically through this endpoint we serve pre-compressed brotli files. The client (browser in this case) is expected to decompress it and consume the data which we managed to do quite easily by explicitly setting the `'Content-Encoding', 'br'` header on the response. This how the code looks like working on 12.3.2

`     

    const [type, data] = await entry.getData()

    if (!type || !data) return res.status(400).end()

    res.setHeader('Content-Type', 'application/json')
    res.setHeader('Content-Encoding', 'br')

    return res.status(200).send(data)

`

However, since our upgrade, we've noticed that this explicit content-encoding header is not being set, and is instead replaced with a `Content-Encoding, gzip` header which causes the UI code to fail as it unexpectedly receives compressed binary data.

I have tried migrating to the new app directory API Handlers, like so: 
`

    const headers = new Headers(req.headers)
    headers.set('Content-Type', 'application/json')
    headers.set('Content-Encoding', 'br')

    return NextResponse.json(data, { status: 200, headers })`

While this does help a little in the sense that the data now comes in the form of a JSON like `{data: <buffer>, type: 'Buffer'}` it still sends the gzip encoding header and I would have to manually decompress my data instead of having the browser do it automatically. 

I have tried disabling NextJS compression through the next.config.js file, but all that does it remove the content-encoding header completely. Is there a way by which I can explicitly set the content-header on my API response?

Thanks,
Daniel

**Comment by HiThere157** [comment, marked as answer]
Hi Daniel,

I am currently facing the same issue and have found a possible workaround until this issue is fixed:

If I set the "Content-Encoding" Header for this particular api endpoint in the [next.config.js configuration](https://nextjs.org/docs/api-reference/next.config.js/headers), it is present in the response:

```ts
/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    appDir: true,
  },
  headers: async () => {
    return [
      {
        source: "/api/compressed_blob/:id",
        headers: [
          {
            key: "Content-Encoding",
            value: "br",
          }
        ]
      },
    ];
  },
  output: "standalone",
};

module.exports = nextConfig;

```

**Reply by Cybertrance** [reply]
Thank you very much! This worked quite well.

**Comment by tapz** [comment]
Is this already fixed? Impossible to use both gzip and brotli if I need to set the Content-Encodiing in the config!

**Comment by tapz** [comment]
I can't believe Next.js has this kind of bugs. This makes it totally unusable in production use.

**Comment by tapz** [comment]
This is a really stupid workaround, but at least it works:

```javascript
      }, {
        source: "/api/foo",
        has: [
          { type: 'header', key: 'Accept-Encoding', value: '*br.*' }
        ],
        headers: [
          {
            key: "Content-Encoding",
            value: "br",
          }
        ]
      }, {
        source: "/api/foo",
        has: [
          { type: 'header', key: 'Accept-Encoding', value: '.*gzip.*)' }
        ],
        headers: [
          {
            key: "Content-Encoding",
            value: "gzip",
          }
        ]
      },
```
